src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,308 @@
+            final Oauth2 oauthRequest = new Oauth2(mAuthRequest, mWebRequestHandler, jwsBuilder);
+            result = oauthRequest.refreshToken(refreshToken);
+            if (result != null && StringExtensions.IsNullOrBlank(result.getRefreshToken())) {
+                Logger.i(TAG, "Refresh token is not returned or empty", "");
+                result.setRefreshToken(refreshToken);
+            }
+        } catch (final IOException | AuthenticationException exc) {
+            // Server side error or similar
+            Logger.e(TAG, "Error in refresh token for request:" + mAuthRequest.getLogInfo(),
+                    ExceptionExtensions.getExceptionMessage(exc), ADALError.AUTH_FAILED_NO_TOKEN,
+                    exc);
+
+            throw new AuthenticationException(
+                    ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
+                    exc);
+        }
+
+        return result;
+    }
+    
+    /**
+     * For testing purpose, inject the dependency. 
+     */
+    void setWebRequestHandler(final IWebRequestHandler webRequestHandler) {
+        this.mWebRequestHandler = webRequestHandler;
+    }
+    
+    /**
+     * Attempt to get new access token with regular RT. 
+     */
+    private AuthenticationResult tryRT() throws AuthenticationException {
+        final TokenCacheItem regularRTItem = mTokenCacheAccessor.getRegularRefreshTokenCacheItem(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), getUserFromRequest());
+        
+        if (regularRTItem == null) {
+            Logger.v(TAG, "Regular token cache entry does not exist, try with MRRT.");
+            return tryMRRT(); 
+        }
+        
+        // When MRRT is returned, we store separate entries for both regular RT entry and MRRT entry, 
+        // we should look for MRRT entry if the token in regular RT entry is marked as MRRT.
+        // However, the current cache implementation never mark the token stored in regular RT entry
+        // as MRRT. To support the backward compatibility and improve cache lookup, when successfully
+        // retrieved regular RT entry token and if the mrrt flag is false, check the existence of MRRT.
+        if (regularRTItem.getIsMultiResourceRefreshToken() || isMRRTEntryExisted()) {
+            final String statusMessage = regularRTItem.getIsMultiResourceRefreshToken() ? 
+                    "Found RT and it's also a MRRT, retry with MRRT" : "RT is found and there is a MRRT entry existed, try with MRRT";
+            Logger.v(TAG, statusMessage);
+            return tryMRRT();
+        }
+        
+        Logger.v(TAG, "Send request to use regular RT for new AT.");
+        return acquireTokenWithCachedItem(regularRTItem);
+    }
+    
+    /**
+     * Attempt to get new access token with MRRT. 
+     * 1) If MRRT does not exist, try with FRT. 
+     * 2) If MRRT is also a FRT, try FRT first.
+     * 3) If MRRT request fails, fall back to FRT.  
```

Comment:
Nope, when MRRT fails, we also try FRT.

---
In reply to: [63268315](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63268315) [](ancestors = 63268315)